samcli/lib/bootstrap/companion_stack/companion_stack_manager.py

```diff
@@ -0,0 +1,265 @@
+            Function logical IDs that need to have ECR repos created
+        """
+        self._builder.clear_functions()
+        for function_logical_id in function_logical_ids:
+            self._builder.add_function(function_logical_id)
+
+    def update_companion_stack(self) -> None:
+        """
+        Blocking call to create or update the companion stack based on current functions
+        Companion stack template will be updated to the s3 bucket first before deployment
+        """
+        if not self._builder.repo_mapping:
+            return
+
+        stack_name = self._companion_stack.stack_name
+        template = self._builder.build()
+
+        with mktempfile() as temporary_file:
+            temporary_file.write(template)
+            temporary_file.flush()
+
+            s3_uploader = S3Uploader(
+                self._s3_client, bucket_name=self._s3_bucket, prefix=self._s3_prefix, no_progressbar=True
+            )
+            # TemplateUrl property requires S3 URL to be in path-style format
+            parts = S3Uploader.parse_s3_url(
+                s3_uploader.upload_with_dedup(temporary_file.name, "template"), version_property="Version"
+            )
+
+        template_url = s3_uploader.to_path_style_s3_url(parts["Key"], parts.get("Version", None))
+
+        exists = self.does_companion_stack_exist()
+        if exists:
+            self._cfn_client.update_stack(
+                StackName=stack_name, TemplateURL=template_url, Capabilities=["CAPABILITY_AUTO_EXPAND"]
+            )
+            waiter = self._cfn_client.get_waiter("stack_update_complete")
+        else:
+            self._cfn_client.create_stack(
+                StackName=stack_name, TemplateURL=template_url, Capabilities=["CAPABILITY_AUTO_EXPAND"]
+            )
+            waiter = self._cfn_client.get_waiter("stack_create_complete")
+
+        waiter.wait(StackName=stack_name, WaiterConfig=self._update_stack_waiter_config)  # type: ignore
+
+    def delete_companion_stack(self):
+        """
+        Blocking call to delete the companion stack
+        """
+        stack_name = self._companion_stack.stack_name
+        waiter = self._cfn_client.get_waiter("stack_delete_complete")
+        self._cfn_client.delete_stack(StackName=stack_name)
+        waiter.wait(StackName=stack_name, WaiterConfig=self._delete_stack_waiter_config)
+
+    def list_deployed_repos(self) -> List[ECRRepo]:
+        """
+        List deployed ECR repos for this companion stack
+        Not using create_change_set as it is slow.
+
+        Returns
+        -------
+        List[ECRRepo]
+            List of ECR repos deployed for this companion stack
+            Returns empty list if companion stack does not exist
+        """
+        if not self.does_companion_stack_exist():
+            return []
+        repos: List[ECRRepo] = list()
+        stack = boto3.resource("cloudformation", config=self._boto_config).Stack(self._companion_stack.stack_name)
```

Comment:
why are we re-creating clients?

Reply:
This is using boto3 resource abstraction instead of doing manual client calls and parsing.